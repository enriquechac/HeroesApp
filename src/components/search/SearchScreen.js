import React, { useMemo } from 'react';
import queryString from 'query-string';

import { useLocation } from 'react-router';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heores/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {
    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);

    const initialForm = {
        searchInput: q,
    };

    const [formValues, handleInputChange] = useForm(initialForm);

    const { searchInput } = formValues;

    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

    const handleSearch = (e) => {
        e.preventDefault();
        // console.log('Search...', searchInput);
        history.push(`?q=${searchInput}`);
    };

    return (
        <div>
            <h1>Search Screen</h1>
            <hr />
            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input
                            type="text"
                            value={searchInput}
                            name="searchInput"
                            placeholder="Find your hero"
                            onChange={handleInputChange}
                            autoComplete="off"
                            className="form-control"
                        />
                        <button
                            type="submit"
                            className="btn w-100 btn-outline-primary my-3"
                        >
                            Search..
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr />
                    {q === '' && (
                        <div className="alert alert-info"> Search a hero </div>
                    )}

                    {q !== '' && heroesFiltered.length === 0 && (
                        <div className="alert alert-info">
                            Hero {q} was not found
                        </div>
                    )}

                    {heroesFiltered.map((hero) => (
                        <HeroCard key={hero.id} {...hero} />
                    ))}
                </div>
            </div>
        </div>
    );
};
