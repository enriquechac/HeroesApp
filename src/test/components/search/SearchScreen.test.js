import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Pruebas en <SearchScreen/>', () => {
    test('Debe de mostrarse correctamente con valores de defecto', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
    });

    test('Debe de mostrar a batman y el input con el valor del query string', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('input').prop('value')).toBe('batman');
    });

    test('Debe de mostrar un herro si no se encuentra el hero', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=error']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();

        expect(wrapper.find('.alert-info').text().trim()).toBe(
            'Hero error was not found'
        );
    });

    test('Debe de llamr el push del history', () => {
        const history = {
            push: jest.fn(),
        };
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=error']}>
                <Route
                    path="/search"
                    component={() => <SearchScreen history={history} />}
                />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'seachText',
                value: 'batman',
            },
        });

        wrapper.find('form').prop('onSubmit')({ preventDefault() {} });

        expect(history.push).toHaveBeenCalledWith('?q=error');
    });
});
