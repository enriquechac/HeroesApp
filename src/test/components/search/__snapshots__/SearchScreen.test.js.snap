// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen/> Debe de mostrar a batman y el input con el valor del query string 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchInput"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn w-100 btn-outline-primary my-3"
          type="submit"
        >
          Search..
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <a
        className="my-card"
        href="/hero/dc-batman"
        onClick={[Function]}
      >
        <img
          alt="Batman"
          className="img img-responsive"
          src="./assets/heroes/dc-batman.jpg"
        />
        <div
          className="profile-name"
        >
          Batman
        </div>
        <div
          className="profile-position"
        >
          Bruce Wayne
        </div>
        <div
          className="profile-overview"
        >
          <div
            className="profile-overview"
          >
            <div
              className="row"
            >
              <div
                className="col-ms-4"
              >
                <h3>
                  DC Comics
                </h3>
                <p>
                  Primera aparici√≥n: 
                  <br />
                  Detective Comics #27
                </p>
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen/> Debe de mostrar un herro si no se encuentra el hero 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchInput"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="error"
        />
        <button
          className="btn w-100 btn-outline-primary my-3"
          type="submit"
        >
          Search..
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Hero 
        error
         was not found
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen/> Debe de mostrarse correctamente con valores de defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchInput"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn w-100 btn-outline-primary my-3"
          type="submit"
        >
          Search..
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
         Search a hero 
      </div>
    </div>
  </div>
</div>
`;
